pallets/randomness/src/lib.rs

```diff
@@ -6,107 +6,127 @@ pub use pallet::*;
+            if let Some(author_vrf) = T::AuthorVrfGetter::get_author_vrf() {
+                AuthorVrf::<T>::put(author_vrf);
+                let current_block = <frame_system::Pallet<T>>::block_number();
+                if AuthorVrfHistory::<T>::count() == HISTORY_SIZE {
+                    AuthorVrfHistory::<T>::remove(current_block - HISTORY_SIZE.into());
+                }
+                AuthorVrfHistory::<T>::insert(current_block, author_vrf);
+            } else {
+                // We don't change the value here, this isn't great but we're not expecting
+                // leader election to fail often enough that it truly affects security.
+                // We're aware this is somewhat wishful thinking but time/output constraints
+                // dictate that this is good enough for now!
+                log::warn!("AuthorVrf is empty, keeping previous value");
+            }
 
-            weight
+            Ok(())
         }
+    }
 
-        fn on_finalize(current_block_number: BlockNumberFor<T>) {
-            // Check if we should clean the seeds
-            if current_block_number % T::CleanupInterval::get() != Zero::zero() {
-                return;
-            }
-
-            // Mark which seeds to remove
-            let mut blocks_to_remove = Vec::new();
-            for creation_height in SeedsMap::<T>::iter_keys() {
-                let age_limit = T::SeedAgeLimit::get();
-                let current_age = current_block_number.saturating_sub(creation_height);
+    #[pallet::inherent]
+    impl<T: Config> ProvideInherent for Pallet<T> {
+        type Call = Call<T>;
+        type Error = InherentError;
+
+        const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+        fn is_inherent_required(_: &InherentData) -> Result<Option<Self::Error>, Self::Error> {
+            // Return Ok(Some(_)) unconditionally because this inherent is required in every block
+            // If it is not found, throw a VrfInherentRequired error.
+            Ok(Some(InherentError::Other(
+                sp_runtime::RuntimeString::Borrowed(
+                    "Inherent required to set babe randomness results",
+                ),
+            )))
+        }
 
-                // Seed is old enough to be removed
-                if current_age >= age_limit {
-                    blocks_to_remove.push(creation_height);
-                }
-            }
+        // The empty-payload inherent extrinsic.
+        fn create_inherent(_data: &InherentData) -> Option<Self::Call> {
+            Some(Call::set_author_vrf {})
+        }
 
-            // Remove old seeds
-            blocks_to_remove.iter().for_each(|number| {
-                SeedsMap::<T>::remove(number);
-            });
+        fn is_inherent(call: &Self::Call) -> bool {
+            matches!(call, Call::set_author_vrf { .. })
         }
     }
 
-    impl<T: Config> Randomness<BlockNumberFor<T>> for Pallet<T> {
```

Comment:
We should be able to fetch the random value for some specific block. That was solved by caching the values for some period.

Reply:
You can:
https://github.com/eigerco/polka-storage/pull/633/files#diff-eface0c86240147709dce013ce11764683db9958eb9a3671ba82034bc2ad0dd4R60-R62

https://github.com/eigerco/polka-storage/pull/633/files#diff-46b4f3deb1d93fe7de8bd7fe5cfbe34abdcf4a8020f4a203a6d5f6a9cd177c5f

Reply:
Aha. Nice.